torchmetrics/functional/text/eed.py

```diff
@@ -0,0 +1,436 @@
+            reference string
+        hyp:
+            hypothesis string
+        alpha:
+            optimal jump penalty, penalty for jumps between characters
+        rho:
+            coverage cost, penalty for repetition of characters
+        deletion:
+            penalty for deletion of character
+        insertion:
+            penalty for insertion or substitution of character
+
+    Return:
+        Extended edit distance score as float
+    """
+    number_of_visits = [-1] * (len(hyp) + 1)
+
+    # row[i] stores cost of cheapest path from (0,0) to (i,l) in CDER aligment grid.
+    row = [1.0] * (len(hyp) + 1)
+
+    row[0] = 0.0  # CDER initialisation 0,0 = 0.0, rest 1.0
+    next_row = [inf] * (len(hyp) + 1)
+
+    for w in range(1, len(ref) + 1):
+        for i in range(0, len(hyp) + 1):
+
+            if i > 0:
+                next_row[i] = min(
+                    next_row[i - 1] + deletion,
+                    row[i - 1] + _distance_between_words(ref[w - 1], hyp[i - 1]),
+                    row[i] + insertion,
+                )
+            else:
+                next_row[i] = row[i] + 1.0
+
+        min_index = next_row.index(min(next_row))
+        number_of_visits[min_index] += 1
+
+        # Long Jumps
+        if ref[w - 1] == " ":
+            jump = alpha + next_row[min_index]
+            next_row = [min(x, jump) for x in next_row]
+
+        row = next_row
+        next_row = [inf] * (len(hyp) + 1)
+
+    coverage = rho * sum(x if x >= 0 else 1 for x in number_of_visits)
+
+    return min(1, (row[-1] + coverage) / (float(len(ref)) + coverage))
+
+
+def _preprocess_en(sentence: str) -> str:
+    """Copied from https://github.com/rwth-i6/ExtendedEditDistance/blob/master/util.py.
+
+    Raises:
+        ValueError: If input sentence is not of a type `str`.
+    """
+    if not isinstance(sentence, str):
+        raise ValueError(f"Only strings allowed during preprocessing step, found {type(sentence)} instead")
+
+    sentence = sentence.rstrip()  # trailing space, tab, or newline
+
+    # Add space before interpunctions
+    rules_interpunction = [
+        (".", " ."),
+        ("!", " !"),
+        ("?", " ?"),
+        (",", " ,"),
+    ]
+    for pattern, replacement in rules_interpunction:
+        sentence = sentence.replace(pattern, replacement)
+
+    rules_re = [
+        (r"\s+", r" "),  # get rid of extra spaces
+        (r"(\d) ([.,]) (\d)", r"\1\2\3"),  # 0 . 1 -> 0.1
+        (r"(Dr|Jr|Prof|Rev|Gen|Mr|Mt|Mrs|Ms) .", r"\1."),  # Mr . -> Mr.
+    ]
+    for pattern, replacement in rules_re:
+        sentence = re.sub(pattern, replacement, sentence)
+
+    # Add space between abbreviations
+    rules_interpunction = [
+        ("e . g .", "e.g."),
+        ("i . e .", "i.e."),
+        ("U . S .", "U.S."),
+    ]
+    for pattern, replacement in rules_interpunction:
+        sentence = sentence.replace(pattern, replacement)
+
+    # add space to beginning of string
+    sentence = " " + sentence
```

Comment:
Both for `"en"` and `"ja"`, there should be added space at the beginning and at the end of a sentence. Also, the final output of pre-processed sentence, after these additions, should not be `sentence` bust `list(sentence)`. I think we handle this a bit differently right now which causes that we obtain different results. See https://github.com/rwth-i6/ExtendedEditDistance/blob/master/EED.py

Reply:
@stancld As mentioned in the paper and in the code, the space is only added to the beginning of the hypothesis and reference sentence, not the end. Also, using lists is inconsequential to the final calculation, which is read-only. Using lists unnecessarily increases time and memory complexity. That said, I'll work on reconciling the results for the original and current implementation.

Reply:
Finally reconciled. The problem was not the string/lists. It was me forgetting to add a space at the end, just like you mentioned.